File: src/main/java/nl/knaw/huygens/alexandria/lmnl/importer/LMNLImporter.java
package nl.knaw.huygens.alexandria.lmnl.importer;

/*
 * #%L
 * alexandria-markup
 * =======
 * Copyright (C) 2016 - 2018 Huygens ING (KNAW)
 * =======
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import nl.knaw.huygens.alexandria.ErrorListener;
import nl.knaw.huygens.alexandria.lmnl.grammar.LMNLLexer;
import nl.knaw.huygens.alexandria.storage.*;
import nl.knaw.huygens.alexandria.storage.wrappers.AnnotationWrapper;
import nl.knaw.huygens.alexandria.storage.wrappers.DocumentWrapper;
import nl.knaw.huygens.alexandria.storage.wrappers.MarkupWrapper;
import nl.knaw.huygens.alexandria.storage.wrappers.TextNodeWrapper;
import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.Token;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.InputStream;
import java.util.*;
import java.util.function.Function;
import java.util.stream.Collectors;

import static java.util.stream.Collectors.joining;

public class LMNLImporter {
  private static final Logger LOG = LoggerFactory.getLogger(LMNLImporter.class);

  private static TAGStore tagStore;

  public LMNLImporter(TAGStore tagStore) {
    LMNLImporter.tagStore = tagStore;
  }

  static class DocumentContext {
    private final TAGDocument document;
    private final Deque<TAGMarkup> openMarkupDeque = new ArrayDeque<>();
    private final Stack<TAGMarkup> openMarkupStack = new Stack<>();
    private final Stack<TAGAnnotation> annotationStack = new Stack<>();
    private final ImporterContext importerContext;

    DocumentContext(TAGDocument document, ImporterContext importerContext) {
      this.document = document;
      this.importerContext = importerContext;
    }

    void openMarkup(TAGMarkup markup) {
      openMarkupDeque.push(markup);
      openMarkupStack.push(markup);
      document.getMarkupIds().add(markup.getId());
    }

    void pushOpenMarkup(String rangeName) {
      // LOG.info("currentDocumentContext().openMarkupDeque={}", openMarkupDeque.stream().map(Markup::getTag).collect(Collectors.toList()));
      Optional<TAGMarkup> findFirst = openMarkupDeque.stream()//
          .filter(tr -> tr.getExtendedTag().equals(rangeName))//
          .findFirst();
      if (findFirst.isPresent()) {
        TAGMarkup markup = findFirst.get();
        if (markup.getTextNodeIds().isEmpty()) {
          // every markup should have at least one textNode
          TAGTextNode emptyTextNode = new TAGTextNode("");
          update(emptyTextNode);
          addTextNode(emptyTextNode);
          closeMarkup();
        }
        openMarkupStack.push(markup);
      } else {
        importerContext.errors.add("Closing tag {" + rangeName + "] found without corresponding open tag.");
      }
    }

    void popOpenMarkup() {
      openMarkupStack.pop();
    }

    void closeMarkup() {
      if (!openMarkupStack.isEmpty()) {
        TAGMarkup markup = openMarkupStack.pop();
        update(markup);
        openMarkupDeque.remove(markup);
      }
    }

    void addTextNode(TAGTextNode textNode) {
      openMarkupDeque.descendingIterator()//
          .forEachRemaining(m -> {
            m.addTextNode(textNode);
            document.associateTextWithMarkup(textNode, m);
          });
      document.addTextNode(textNode);
    }

    private TAGMarkup currentMarkup() {
      return openMarkupDeque.isEmpty() ? null : openMarkupStack.peek();
    }

    void openAnnotation(TAGAnnotation annotation) {
      if (annotationStack.isEmpty()) {
        TAGMarkup markup = currentMarkup();
        if (markup != null) {
          markup.addAnnotation(annotation);
        }
      } else {
        annotationStack.peek().addAnnotation(annotation);
      }
      annotationStack.push(annotation);
    }

    TAGDocument currentAnnotationDocument() {
      Long value = annotationStack.peek().getDocumentId();
      return tagStore.getDocument(value);
    }

    void closeAnnotation() {
      TAGAnnotation annotation = annotationStack.pop();
      update(annotation);
    }
  }

  static class ImporterContext {
    private final Deque<DocumentContext> documentContextStack = new ArrayDeque<>();
    private final LMNLLexer lexer;
    private final List<String> errors = new ArrayList<>();

    ImporterContext(LMNLLexer lexer) {
      this.lexer = lexer;
    }

    Token nextToken() {
      return lexer.nextToken();
    }

    String getModeName() {
      return lexer.getModeNames()[lexer._mode];
    }

    String getRuleName() {
      return lexer.getRuleNames()[lexer.getToken().getType() - 1];
    }

    void pushDocumentContext(TAGDocument document) {
      documentContextStack.push(new DocumentContext(document, this));
    }

    DocumentContext currentDocumentContext() {
      return documentContextStack.peek();
    }

    DocumentContext popDocumentContext() {
      DocumentContext documentContext = documentContextStack.pop();
      update(documentContext.document);
      if (!documentContext.openMarkupDeque.isEmpty()) {
        String openRanges = documentContext.openMarkupDeque.stream()//
            .map(m -> "[" + m.getExtendedTag() + "}")//
            .collect(Collectors.joining(", "));
        errors.add("Unclosed LMNL range(s): " + openRanges);
      }
      return documentContext;
    }

    TAGMarkup newMarkup(String tagName) {
      TAGMarkup tagMarkup = new TAGMarkup(currentDocumentContext().document.getId(), tagName);
      update(tagMarkup);
      return tagMarkup;
    }

    void openMarkup(TAGMarkup markup) {
      currentDocumentContext().openMarkup(markup);
    }

    void pushOpenMarkup(String rangeName) {
      currentDocumentContext().pushOpenMarkup(rangeName);
    }

    void popOpenMarkup() {
      currentDocumentContext().popOpenMarkup();
    }

    void closeMarkup() {
      currentDocumentContext().closeMarkup();
    }

    void addTextNode(TAGTextNode textNode) {
      currentDocumentContext().addTextNode(textNode);
    }

    void openAnnotation(TAGAnnotation annotation) {
      currentDocumentContext().openAnnotation(annotation);
    }

    TAGDocument currentAnnotationDocument() {
      return currentDocumentContext().currentAnnotationDocument();
    }

    void closeAnnotation() {
      currentDocumentContext().closeAnnotation();
    }

    List<String> getErrors() {
      return errors;
    }

    boolean hasErrors() {
      return !errors.isEmpty();
    }
  }

  public DocumentWrapper importLMNL(String input) throws LMNLSyntaxError {
    CharStream antlrInputStream = CharStreams.fromString(input);
    return importLMNL(antlrInputStream);
  }

  public DocumentWrapper importLMNL(InputStream input) throws LMNLSyntaxError {
    try {
      CharStream antlrInputStream = CharStreams.fromStream(input);
      return importLMNL(antlrInputStream);
    } catch (IOException e) {
      e.printStackTrace();
      throw new RuntimeException(e);
    }
  }

  private DocumentWrapper importLMNL(CharStream antlrInputStream) throws LMNLSyntaxError {
    LMNLLexer lexer = new LMNLLexer(antlrInputStream);
    ErrorListener errorListener = new ErrorListener();
    lexer.addErrorListener(errorListener);

    ImporterContext context = new ImporterContext(lexer);
    DocumentWrapper documentWrapper = tagStore.createDocumentWrapper();
    TAGDocument document = documentWrapper.getDocument();
    update(document);
    context.pushDocumentContext(document);
    handleDefaultMode(context);
    joinDiscontinuedRanges(documentWrapper);
    context.popDocumentContext();

    String errorMsg = "";
    if (context.hasErrors()) {
      String errors = context.getErrors().stream().collect(Collectors.joining("\n"));
      errorMsg = "Parsing errors:\n" + errors;
    }
    if (errorListener.hasErrors()) {
      String errors = errorListener.getErrors().stream().collect(Collectors.joining("\n"));
      errorMsg += "\n\nTokenizing errors:\n" + errors;
    }
    if (!errorMsg.isEmpty()) {
      throw new LMNLSyntaxError(errorMsg);
    }
    update(document);
    return documentWrapper;
  }

  private void handleDefaultMode(ImporterContext context) {
    String methodName = "defaultMode";
    Token token;
    do {
      token = context.nextToken();
      if (token.getType() != Token.EOF) {
        String ruleName = context.getRuleName();
        String modeName = context.getModeName();
        log(methodName, ruleName, modeName, token, context);
        switch (token.getType()) {
          case LMNLLexer.BEGIN_OPEN_RANGE:
            handleOpenRange(context);
            break;

          case LMNLLexer.BEGIN_CLOSE_RANGE:
            handleCloseRange(context);
            break;

          case LMNLLexer.TEXT:
            TAGTextNode textNode = new TAGTextNode(token.getText());
            update(textNode);
            context.addTextNode(textNode);
            break;

          // case LMNLLexer.TagOpenStartChar:
          // case LMNLLexer.TagOpenEndChar:
          // case LMNLLexer.TagCloseStartChar:
          // case LMNLLexer.TagCloseEndChar:
          // break;

          default:
            handleUnexpectedToken(methodName, token, ruleName, modeName);
            break;
        }
      }
    } while (token.getType() != Token.EOF);
  }


  private void handleOpenRange(ImporterContext context) {
    String methodName = "handleOpenRange";
    boolean goOn = true;
    while (goOn) {
      Token token = context.nextToken();
      String ruleName = context.getRuleName();
      String modeName = context.getModeName();
      log(methodName, ruleName, modeName, token, context);
      switch (token.getType()) {
        case LMNLLexer.Name_Open_Range:
          TAGMarkup markup = context.newMarkup(token.getText());
          context.openMarkup(markup);
          break;
        case LMNLLexer.BEGIN_OPEN_ANNO:
          handleAnnotation(context);
          break;
        case LMNLLexer.END_OPEN_RANGE:
          context.popOpenMarkup();
          goOn = false;
          break;
        case LMNLLexer.END_ANONYMOUS_RANGE:
          TAGTextNode textNode = new TAGTextNode("");
          update(textNode);
          context.addTextNode(textNode);
          context.closeMarkup();
          goOn = false;
          break;

        // case LMNLLexer.TagOpenStartChar:
        // case LMNLLexer.TagOpenEndChar:
        // case LMNLLexer.TagCloseStartChar:
        // case LMNLLexer.TagCloseEndChar:
        // break;

        default:
          handleUnexpectedToken(methodName, token, ruleName, modeName);
          break;
      }
      goOn = goOn && token.getType() != Token.EOF;
    }
  }

  private void handleAnnotation(ImporterContext context) {
    String methodName = "handleAnnotation";
    TAGAnnotation annotation = tagStore.createAnnotation("");
    context.openAnnotation(annotation);
    boolean goOn = true;
    while (goOn) {
      Token token = context.nextToken();
      String ruleName = context.getRuleName();
      String modeName = context.getModeName();
      log(methodName, ruleName, modeName, token, context);
      switch (token.getType()) {
        case LMNLLexer.Name_Open_Annotation:
          annotation.setTag(token.getText());
          break;
        case LMNLLexer.OPEN_ANNO_IN_ANNO_OPENER:
        case LMNLLexer.OPEN_ANNO_IN_ANNO_CLOSER:
          handleAnnotation(context);
          break;
        case LMNLLexer.END_OPEN_ANNO:
          context.pushDocumentContext(context.currentAnnotationDocument());
          break;

        case LMNLLexer.ANNO_TEXT:
          TAGTextNode textNode = new TAGTextNode(token.getText());
          update(textNode);
          context.addTextNode(textNode);
          break;

        case LMNLLexer.BEGIN_ANNO_OPEN_RANGE:
          handleOpenRange(context);
          break;

        case LMNLLexer.BEGIN_ANNO_CLOSE_RANGE:
          handleCloseRange(context);
          break;

        case LMNLLexer.BEGIN_CLOSE_ANNO:
        case LMNLLexer.Name_Close_Annotation:
          break;
        case LMNLLexer.END_CLOSE_ANNO:
          context.popDocumentContext();
        case LMNLLexer.END_EMPTY_ANNO:
          context.closeAnnotation();
          goOn = false;
          break;

        // case LMNLLexer.TagOpenStartChar:
        // case LMNLLexer.TagOpenEndChar:
        // case LMNLLexer.TagCloseStartChar:
        // case LMNLLexer.TagCloseEndChar:
        // break;

        default:
          handleUnexpectedToken(methodName, token, ruleName, modeName);
          break;
      }
      goOn = goOn && token.getType() != Token.EOF;
    }
  }

  private void handleCloseRange(ImporterContext context) {
    String methodName = "handleCloseRange";
    boolean goOn = true;
    while (goOn) {
      Token token = context.nextToken();
      String ruleName = context.getRuleName();
      String modeName = context.getModeName();
      log(methodName, ruleName, modeName, token, context);
      switch (token.getType()) {
        case LMNLLexer.Name_Close_Range:
          String rangeName = token.getText();
          context.pushOpenMarkup(rangeName);
          break;
        case LMNLLexer.BEGIN_OPEN_ANNO_IN_RANGE_CLOSER:
          handleAnnotation(context);
          break;
        case LMNLLexer.END_CLOSE_RANGE:
          context.closeMarkup();
          goOn = false;
          break;

        // case LMNLLexer.TagOpenStartChar:
        // case LMNLLexer.TagOpenEndChar:
        // case LMNLLexer.TagCloseStartChar:
        // case LMNLLexer.TagCloseEndChar:
        // break;

        default:
          handleUnexpectedToken(methodName, token, ruleName, modeName);
          break;
      }
      goOn = goOn && token.getType() != Token.EOF;
    }
  }

  private void handleUnexpectedToken(String methodName, Token token, String ruleName, String modeName) {
    String message = methodName + ": unexpected rule/token: token=" + token + ", ruleName=" + ruleName + ", mode=" + modeName;
    LOG.error(message);
    throw new LMNLSyntaxError(message);
  }

//  private static void joinDiscontinuedRanges(Document document) {
//    joinDiscontinuedRanges(document.getDocumentId());
//  }

  private static void joinDiscontinuedRanges(DocumentWrapper document) {
    Map<String, TAGMarkup> markupsToJoin = new HashMap<>();
    List<Long> markupIdsToRemove = new ArrayList<>();
    document.getMarkupStream()//
        .filter(MarkupWrapper::hasN)//
        .forEach(markup -> {
          String tag = markup.getTag();
          AnnotationWrapper annotation = markup.getAnnotationStream()//
              .filter(a -> a.getTag().equals("n"))//
              .findFirst()//
              .get();
          String key = tag + "-" + annotationText(annotation);
          if (markupsToJoin.containsKey(key)) {
            TAGMarkup originalMarkup = markupsToJoin.get(key);
            markup.getMarkup().getAnnotationIds().remove(annotation.getId());
            document.joinMarkup(originalMarkup, markup);
            markupIdsToRemove.add(markup.getId());
          } else {
            markupsToJoin.put(key, markup.getMarkup());
          }
        });

    document.getDocument().getMarkupIds().removeAll(markupIdsToRemove);
    document.getMarkupStream()//
        .map(MarkupWrapper::getAnnotationStream)//
        .flatMap(Function.identity())//
        .map(AnnotationWrapper::getDocument)//
        .forEach(LMNLImporter::joinDiscontinuedRanges);
  }

  private static String annotationText(AnnotationWrapper annotation) {
    return annotation.getDocument().getTextNodeStream()//
        .map(TextNodeWrapper::getText)//
        .collect(joining());
  }

  private void log(String mode, String ruleName, String modeName, Token token, ImporterContext context) {
    // LOG.info("{}:\tlevel:{}, <{}> :\t{} ->\t{}", //
    // mode, context.limenContextStack.size(), //
    // token.getText().replace("\n", "\\n"), //
    // ruleName, modeName);
  }

  private static Long update(TAGObject tagObject) {
    return tagStore.persist(tagObject);
  }

}

File: main/src/main/kotlin/nl/knaw/huygens/alexandria/lmnl/importer/LMNLImporter.kt
package nl.knaw.huygens.alexandria.lmnl.importer

/*
 * #%L
 * alexandria-markup-core
 * =======
 * Copyright (C) 2016 - 2020 HuC DI (KNAW)
 * =======
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import nl.knaw.huygens.alexandria.ErrorListener
import nl.knaw.huygens.alexandria.lmnl.grammar.LMNLLexer
import nl.knaw.huygens.alexandria.storage.TAGAnnotation
import nl.knaw.huygens.alexandria.storage.TAGDocument
import nl.knaw.huygens.alexandria.storage.TAGMarkup
import nl.knaw.huygens.alexandria.storage.TAGStore
import nl.knaw.huygens.alexandria.storage.dto.*
import org.antlr.v4.runtime.CharStream
import org.antlr.v4.runtime.CharStreams
import org.antlr.v4.runtime.Token
import org.slf4j.LoggerFactory
import java.io.IOException
import java.io.InputStream
import java.io.UncheckedIOException
import java.util.*
import java.util.stream.Collectors

class LMNLImporter(tagStore: TAGStore?) {
  @Throws(LMNLSyntaxError::class)
  fun importLMNL(input: String?): TAGDocument {
    val antlrInputStream: CharStream = CharStreams.fromString(input)
    return importLMNL(antlrInputStream)
  }

  @Throws(LMNLSyntaxError::class)
  fun importLMNL(input: InputStream?): TAGDocument {
    return try {
      val antlrInputStream = CharStreams.fromStream(input)
      importLMNL(antlrInputStream)
    } catch (e: IOException) {
      e.printStackTrace()
      throw UncheckedIOException(e)
    }
  }

  @Throws(LMNLSyntaxError::class)
  private fun importLMNL(antlrInputStream: CharStream): TAGDocument {
    val lexer = LMNLLexer(antlrInputStream)
    val errorListener = ErrorListener()
    lexer.addErrorListener(errorListener)
    val context = ImporterContext(lexer)
    val document = tagStore!!.createDocument()
    val dto = document.dto
    update(dto)
    context.pushDocumentContext(dto)
    handleDefaultMode(context)
    joinDiscontinuedRanges(document)
    context.popDocumentContext()
    var errorMsg = ""
    if (context.hasErrors()) {
      val errors = java.lang.String.join("\n", context.getErrors())
      errorMsg = "Parsing errors:\n$errors"
    }
    if (errorListener.hasErrors()) {
      errorMsg += """
        
        
        Tokenizing errors:
        ${errorListener.prefixedErrorMessagesAsString}
        """.trimIndent()
    }
    if (errorMsg.isNotEmpty()) {
      throw LMNLSyntaxError(errorMsg)
    }
    update(dto)
    return document
  }

  private fun handleDefaultMode(context: ImporterContext) {
    val methodName = "defaultMode"
    var token: Token
    do {
      token = context.nextToken()
      if (token.type != Token.EOF) {
        val ruleName = context.ruleName
        val modeName = context.modeName
        log(methodName, ruleName, modeName, token, context)
        when (token.type) {
          LMNLLexer.BEGIN_OPEN_RANGE -> handleOpenRange(context)
          LMNLLexer.BEGIN_CLOSE_RANGE -> handleCloseRange(context)
          LMNLLexer.TEXT -> {
            val textNode = TAGTextNodeDTO(token.text)
            update(textNode)
            context.addTextNode(textNode)
          }
          else -> handleUnexpectedToken(methodName, token, ruleName, modeName)
        }
      }
    } while (token.type != Token.EOF)
  }

  private fun handleOpenRange(context: ImporterContext) {
    val methodName = "handleOpenRange"
    var goOn = true
    while (goOn) {
      val token = context.nextToken()
      val ruleName = context.ruleName
      val modeName = context.modeName
      log(methodName, ruleName, modeName, token, context)
      when (token.type) {
        LMNLLexer.Name_Open_Range -> {
          val markup = context.newMarkup(token.text)
          context.openMarkup(markup)
        }
        LMNLLexer.BEGIN_OPEN_ANNO -> handleAnnotation(context)
        LMNLLexer.END_OPEN_RANGE -> {
          context.popOpenMarkup()
          goOn = false
        }
        LMNLLexer.END_ANONYMOUS_RANGE -> {
          val textNode = TAGTextNodeDTO("")
          update(textNode)
          context.addTextNode(textNode)
          context.closeMarkup()
          goOn = false
        }
        else -> handleUnexpectedToken(methodName, token, ruleName, modeName)
      }
      goOn = goOn && token.type != Token.EOF
    }
  }

  private fun handleAnnotation(context: ImporterContext) {
    val methodName = "handleAnnotation"
    //    TAGAnnotationDTO annotation = tagStore.createAnnotationDTO("");
    //    context.openAnnotation(annotation);
    //    boolean goOn = true;
    //    while (goOn) {
    //      Token token = context.nextToken();
    //      String ruleName = context.getRuleName();
    //      String modeName = context.getModeName();
    //      log(methodName, ruleName, modeName, token, context);
    //      switch (token.getType()) {
    //        case LMNLLexer.Name_Open_Annotation:
    //          annotation.setKey(token.getText());
    //          break;
    //        case LMNLLexer.OPEN_ANNO_IN_ANNO_OPENER:
    //        case LMNLLexer.OPEN_ANNO_IN_ANNO_CLOSER:
    //          handleAnnotation(context);
    //          break;
    //        case LMNLLexer.END_OPEN_ANNO:
    //          context.pushDocumentContext(context.currentAnnotationDocument());
    //          break;

    //        case LMNLLexer.ANNO_TEXT:
    //          TAGTextNodeDTO textNode = new TAGTextNodeDTO(token.getText());
    //          update(textNode);
    //          context.addTextNode(textNode);
    //          break;

    //        case LMNLLexer.BEGIN_ANNO_OPEN_RANGE:
    //          handleOpenRange(context);
    //          break;

    //        case LMNLLexer.BEGIN_ANNO_CLOSE_RANGE:
    //          handleCloseRange(context);
    //          break;

    //        case LMNLLexer.BEGIN_CLOSE_ANNO:
    //        case LMNLLexer.Name_Close_Annotation:
    //          break;
    //        case LMNLLexer.END_CLOSE_ANNO:
    //          context.popDocumentContext();
    //        case LMNLLexer.END_EMPTY_ANNO:
    //          context.closeAnnotation();
    //          goOn = false;
    //          break;

    //        // case LMNLLexer.TagOpenStartChar:
    //        // case LMNLLexer.TagOpenEndChar:
    //        // case LMNLLexer.TagCloseStartChar:
    //        // case LMNLLexer.TagCloseEndChar:
    //        // break;

    //        default:
    //          handleUnexpectedToken(methodName, token, ruleName, modeName);
    //          break;
    //      }
    //      goOn = goOn && token.getType() != Token.EOF;
    //    }
  }

  private fun handleCloseRange(context: ImporterContext) {
    val methodName = "handleCloseRange"
    var goOn = true
    while (goOn) {
      val token = context.nextToken()
      val ruleName = context.ruleName
      val modeName = context.modeName
      log(methodName, ruleName, modeName, token, context)
      when (token.type) {
        LMNLLexer.Name_Close_Range -> {
          val rangeName = token.text
          context.pushOpenMarkup(rangeName)
        }
        LMNLLexer.BEGIN_OPEN_ANNO_IN_RANGE_CLOSER -> handleAnnotation(context)
        LMNLLexer.END_CLOSE_RANGE -> {
          context.closeMarkup()
          goOn = false
        }
        else -> handleUnexpectedToken(methodName, token, ruleName, modeName)
      }
      goOn = goOn && token.type != Token.EOF
    }
  }

  private fun handleUnexpectedToken(
      methodName: String, token: Token, ruleName: String, modeName: String) {
    val message = (methodName
        + ": unexpected rule/token: token="
        + token
        + ", ruleName="
        + ruleName
        + ", mode="
        + modeName)
    LOG.error(message)
    throw LMNLSyntaxError(message)
  }

  //  private static void joinDiscontinuedRanges(Document document) {
  //    joinDiscontinuedRanges(document.getDocumentId());
  //  }
  private fun log(
      mode: String, ruleName: String, modeName: String, token: Token, context: ImporterContext) {
    // LOG.info("{}:\tlevel:{}, <{}> :\t{} ->\t{}",
    // mode, context.limenContextStack.size(),
    // token.getText().replace("\n", "\\n"),
    // ruleName, modeName);
  }

  internal class DocumentContext(val document: TAGDocumentDTO, private val importerContext: ImporterContext) {
    val openMarkupDeque: Deque<TAGMarkupDTO> = ArrayDeque()
    private val openMarkupStack = Stack<TAGMarkupDTO>()
    private val annotationStack = Stack<TAGAnnotationDTO>()
    fun openMarkup(markup: TAGMarkupDTO) {
      openMarkupDeque.push(markup)
      openMarkupStack.push(markup)
      document.markupIds.add(markup.dbId)
    }

    fun pushOpenMarkup(rangeName: String) {
      // LOG.info("currentDocumentContext().openMarkupDeque={}",
      // openMarkupDeque.stream().map(Markup::getKey).collect(Collectors.toList()));
      val findFirst = openMarkupDeque.stream()
          .map { dto: TAGMarkupDTO? -> TAGMarkup(tagStore, dto) }
          .filter { m: TAGMarkup -> m.extendedTag == rangeName }
          .findFirst()
      if (findFirst.isPresent) {
        val markup = findFirst.get()
        if (!document.markupHasTextNodes(markup)) {
          // every markup should have at least one textNode
          val emptyTextNode = TAGTextNodeDTO("")
          update(emptyTextNode)
          addTextNode(emptyTextNode)
          closeMarkup()
        }
        openMarkupStack.push(markup.dto)
      } else {
        importerContext.getErrors().add(
            "Closing tag {$rangeName] found without corresponding open tag.")
      }
    }

    fun popOpenMarkup() {
      openMarkupStack.pop()
    }

    fun closeMarkup() {
      if (!openMarkupStack.isEmpty()) {
        val markup = openMarkupStack.pop()
        update(markup)
        openMarkupDeque.remove(markup)
      }
    }

    fun addTextNode(textNode: TAGTextNodeDTO) {
      openMarkupDeque
          .descendingIterator()
          .forEachRemaining { m: TAGMarkupDTO? ->
            //            m.addTextNode(textNode);
            document.associateTextWithMarkupForLayer(textNode, m, "")
          }
      if (!document.hasTextNodes()) {
        document.firstTextNodeId = textNode.dbId
      }
      document.addTextNode(textNode)
    }

    private fun currentMarkup(): TAGMarkupDTO? {
      return if (openMarkupDeque.isEmpty()) null else openMarkupStack.peek()
    }

    fun openAnnotation(annotation: TAGAnnotationDTO) {
      if (annotationStack.isEmpty()) {
        val markup = currentMarkup()
        if (markup != null) {
          //          markup.addAnnotation(annotation);
        }
      } else {
        //        annotationStack.peek().addAnnotation(annotation);
      }
      annotationStack.push(annotation)
    }

    fun currentAnnotationDocument(): TAGDocumentDTO? {
      return null
      //      Long value = annotationStack.peek().getDocumentId();
      //      return tagStore.getDocumentDTO(value);
    }

    fun closeAnnotation() {
      val annotation = annotationStack.pop()
      update(annotation)
    }

  }

  internal class ImporterContext(private val lexer: LMNLLexer) {
    private val documentContextStack: Deque<DocumentContext> = ArrayDeque()
    private val _errors: MutableList<String> = ArrayList()

    fun nextToken(): Token {
      return lexer.nextToken()
    }

    val modeName: String
      get() = lexer.modeNames[lexer._mode]

    val ruleName: String
      get() = lexer.ruleNames[lexer.token.type - 1]

    fun pushDocumentContext(document: TAGDocumentDTO) {
      documentContextStack.push(DocumentContext(document, this))
    }

    fun currentDocumentContext(): DocumentContext {
      return documentContextStack.peek()
    }

    fun popDocumentContext(): DocumentContext {
      val documentContext = documentContextStack.pop()
      update(documentContext.document)
      if (!documentContext.openMarkupDeque.isEmpty()) {
        val openRanges = documentContext.openMarkupDeque.stream()
            .map { dto: TAGMarkupDTO? -> TAGMarkup(tagStore, dto) }
            .map { m: TAGMarkup -> "[" + m.extendedTag + "}" }
            .collect(Collectors.joining(", "))
        _errors.add("Unclosed LMNL range(s): $openRanges")
      }
      return documentContext
    }

    fun newMarkup(tagName: String?): TAGMarkupDTO {
      val tagMarkupDTO = TAGMarkupDTO(currentDocumentContext().document.dbId, tagName)
      update(tagMarkupDTO)
      return tagMarkupDTO
    }

    fun openMarkup(markup: TAGMarkupDTO) {
      currentDocumentContext().openMarkup(markup)
    }

    fun pushOpenMarkup(rangeName: String) {
      currentDocumentContext().pushOpenMarkup(rangeName)
    }

    fun popOpenMarkup() {
      currentDocumentContext().popOpenMarkup()
    }

    fun closeMarkup() {
      currentDocumentContext().closeMarkup()
    }

    fun addTextNode(textNode: TAGTextNodeDTO) {
      currentDocumentContext().addTextNode(textNode)
    }

    fun openAnnotation(annotation: TAGAnnotationDTO) {
      currentDocumentContext().openAnnotation(annotation)
    }

    fun currentAnnotationDocument(): TAGDocumentDTO? {
      return currentDocumentContext().currentAnnotationDocument()
    }

    fun closeAnnotation() {
      currentDocumentContext().closeAnnotation()
    }

    fun getErrors(): MutableList<String> {
      return _errors
    }

    fun hasErrors(): Boolean {
      return _errors.isNotEmpty()
    }

  }

  companion object {
    private val LOG = LoggerFactory.getLogger(LMNLImporter::class.java)
    private var tagStore: TAGStore? = null
    private fun joinDiscontinuedRanges(document: TAGDocument) {
      //    Map<String, TAGMarkupDTO> markupsToJoin = new HashMap<>();
      //    List<Long> markupIdsToRemove = new ArrayList<>();
      //    document.getMarkupStream()
      //        .filter(TAGMarkup::hasN)
      //        .forEach(markup -> {
      //          String tag = markup.getTag();
      //          AnnotationInfo annotation = markup.getAnnotationStream()
      //              .filter(a -> a.getName().equals("n"))
      //              .findFirst()
      //              .get();
      //          String key = tag + "-" + annotationText(annotation);
      //          if (markupsToJoin.containsKey(key)) {
      //            TAGMarkupDTO originalMarkup = markupsToJoin.get(key);
      //            markup.getDTO().getAnnotationIds().remove(annotation.getResourceId());
      ////            document.joinMarkup(originalMarkup, markup);
      //            markupIdsToRemove.add(markup.getResourceId());
      //          } else {
      //            markupsToJoin.put(key, markup.getDTO());
      //          }
      //        });

      //    document.getDTO().getMarkupIds().removeAll(markupIdsToRemove);
      ////    document.getMarkupStream()
      ////        .map(TAGMarkup::getAnnotationStream)
      ////        .flatMap(Function.identity())
      ////        .map(TAGAnnotation::getDocument)
      ////        .forEach(LMNLImporter::joinDiscontinuedRanges);
    }

    private fun annotationText(annotation: TAGAnnotation): String {
      return "TODO"
      //    return annotation.getDocument().getTextNodeStream()
      //        .map(TAGTextNode::getText)
      //        .collect(joining());
    }

    private fun update(tagdto: TAGDTO): Long {
      return tagStore!!.persist(tagdto)
    }
  }

  init {
    Companion.tagStore = tagStore
  }
}
